### PR TITLE
Statistics caching update from pickle to safetensors

### PR DESCRIPTION
**Pull request by kshpv**
### Changes

Update module to dump and load statistics from pickle to safetensors.
Statistics are saved and are loaded in the backend tensor representation (torch, numpy).

### Reason for changes

Safe issues with pickle

### Related tickets

157343

### Tests

Tests were updated

### CODE REVIEW

**Comment by kshpv** [comment]
ptq weight compression 269
test_examples 633

File: nncf/experimental/common/tensor_statistics/statistics.py [hunk 1]
```diff
@@ -252,6 +314,17 @@ def __eq__(self, other: Any) -> bool:
...
 
+    def get_data_to_dump(self):
+        tensor_method = get_tensor_method(self.mean_values[0].backend)
+        return {
+            self.MEAN_STAT: tensor_method([mean_value.data for mean_value in self.mean_values]),
+            self.SHAPE_STAT: tensor_method([[dim.data for dim in shape] for shape in self.shape_values]),
+        }
+
+    def load_data(self, loaded_data: Dict[str, TTensor]) -> None:
+        self.shape_values = [tuple(shape.tolist()) for shape in loaded_data[self.SHAPE_STAT]]
```
Comment (AlexanderDokuchaev) [inline nncf/experimental/common/tensor_statistics/statistics.py:325]: Direct call .tolist() method for any tensor type
Looks like it should be done like `Tensor(loaded_data[self.SHAPE_STAT]).tolist()`

File: nncf/experimental/common/tensor_statistics/statistics.py [hunk 2]
```diff
@@ -237,7 +299,7 @@ class WCTensorStatistic(TensorStatistic):
     SHAPE_STAT = "shape_values"
 
     mean_values: List[Tensor]
-    shape_values: List[Tuple[int, ...]]
+    shape_values: List[Tuple[Tensor]]
```
Comment (AlexanderDokuchaev) [inline nncf/experimental/common/tensor_statistics/statistics.py:302]: ?
`        self.shape_values = [tuple(shape.tolist()) for shape in loaded_data[self.SHAPE_STAT]]`

File: nncf/common/tensor_statistics/statistics_serializer.py [hunk 3]
```diff
@@ -33,87 +35,128 @@ def load_metadata(dir_path: Path) -> Dict[str, Any]:
...
-                pickle.dump(statistics_value, f)
-        except (IOError, pickle.PicklingError) as e:
+            save_file_func(statistics_value, file_path)
+        except Exception as e:
             raise nncf.InternalError(f"Failed to write data to file {file_path}: {e}")
 
-    # Add additional metadata if provided
     if additional_metadata:
-        metadata["metadata"] = additional_metadata
+        metadata |= additional_metadata
```
Comment (AlexanderDokuchaev) [inline nncf/common/tensor_statistics/statistics_serializer.py:120]: Where i can find description of metadata structure?

**Review by AlexanderDokuchaev — CHANGES_REQUESTED** [review]

File: nncf/common/tensor_statistics/statistics_serializer.py [hunk 4]
```diff
@@ -8,23 +8,25 @@
...
 from nncf.common.utils.os import safe_open
+from nncf.tensor.definitions import TensorBackend
+from nncf.tensor.tensor import TTensor
 
 METADATA_FILE = "statistics_metadata.json"
 
 
 def sanitize_filename(filename: str) -> str:
     """
     Replaces any forbidden characters with an underscore.
```
[the remark this answers is not in the document]
Comment (kshpv) [reply nncf/common/tensor_statistics/statistics_serializer.py:42]: done

File: nncf/common/tensor_statistics/statistics_serializer.py [hunk 5]
```diff
@@ -33,87 +35,128 @@ def load_metadata(dir_path: Path) -> Dict[str, Any]:
...
+    :param dir_path: Path to the directory where the metadata file will be saved.
     """
     metadata_file = dir_path / METADATA_FILE
     with safe_open(metadata_file, "w") as f:
-        json.dump(metadata, f, indent=4)
+        json.dump(metadata, cast(TextIO, f), indent=4)
 
 
-def load_from_dir(dir_path: str) -> Tuple[Dict[str, Any], Dict[str, str]]:
+def load_from_dir(dir_path: str, backend: TensorBackend) -> Tuple[Dict[str, Dict[str, TTensor]], Dict[str, Any]]:
```
[the remark this answers is not in the document]
Comment (kshpv) [reply nncf/common/tensor_statistics/statistics_serializer.py]: done

File: nncf/common/tensor_statistics/statistics_serializer.py [hunk 6]
```diff
@@ -33,87 +35,128 @@ def load_metadata(dir_path: Path) -> Dict[str, Any]:
...
     :param dir_path: The directory where the metadata file is stored.
-    :return: A dictionary containing the mapping and metadata.
+    :return: A dictionary containing the metadata.
     """
     metadata_file = dir_path / METADATA_FILE
     if metadata_file.exists():
         with safe_open(metadata_file, "r") as f:
             return cast(Dict[str, Any], json.load(f))
-    return {"mapping": {}, "metadata": {}}
+    return {"mapping": {}}
```
[the remark this answers is not in the document]
Comment (kshpv) [reply nncf/common/tensor_statistics/statistics_serializer.py]: good suggestion. I added checks on this behavior covered by tests

File: nncf/common/tensor_statistics/statistics_serializer.py [hunk 7]
```diff
@@ -33,87 +35,128 @@ def load_metadata(dir_path: Path) -> Dict[str, Any]:
...
+
+def get_save_file_method(tensor_backend: TensorBackend) -> Callable[..., Any]:
+    """
+    Returns the appropriate save_file function based on the backend.
+
+    :param tensor_backend: Tensor backend type.
+    :return: Function to save tensors.
+    """
+    try:
+        if tensor_backend == TensorBackend.numpy:
```
[the remark this answers is not in the document]
Comment (kshpv) [reply nncf/common/tensor_statistics/statistics_serializer.py]: done

File: nncf/experimental/common/tensor_statistics/statistics.py [hunk 8]
```diff
@@ -252,6 +314,17 @@ def __eq__(self, other: Any) -> bool:
         )
         return mean_values_equal
 
+    def get_data_to_dump(self):
+        tensor_method = get_tensor_method(self.mean_values[0].backend)
+        return {
+            self.MEAN_STAT: tensor_method([mean_value.data for mean_value in self.mean_values]),
```
[the remark this answers is not in the document]
Comment (kshpv) [reply nncf/experimental/common/tensor_statistics/statistics.py]: done

File: nncf/experimental/common/tensor_statistics/statistics.py (hunk 1 above)
Comment (kshpv) [reply nncf/experimental/common/tensor_statistics/statistics.py]: updated. Please check

File: nncf/experimental/common/tensor_statistics/statistics.py (hunk 2 above)
Comment (kshpv) [reply nncf/experimental/common/tensor_statistics/statistics.py:305]: done

File: nncf/common/tensor_statistics/statistics_serializer.py (hunk 3 above)
Comment (kshpv) [reply nncf/common/tensor_statistics/statistics_serializer.py:151]: Added a small description at the file begginig

File: nncf/common/tensor_statistics/statistics_serializer.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (kshpv) [reply nncf/common/tensor_statistics/statistics_serializer.py]: added discpatcing on Tensor lvl. please check

File: nncf/common/tensor_statistics/statistics_serializer.py [hunk 9]
```diff
@@ -8,112 +8,141 @@
...
 import nncf
-from nncf.common.utils.os import fail_if_symlink
 from nncf.common.utils.os import safe_open
+from nncf.tensor import functions as fns
+from nncf.tensor.definitions import TensorBackend
+from nncf.tensor.tensor import Tensor
 
 METADATA_FILE = "statistics_metadata.json"
 
+# - Metadata is stored in a JSON file named "statistics_metadata.json".
```
Comment (alexsu52) [inline nncf/common/tensor_statistics/statistics_serializer.py:25]: Could you mention why you need to use metadata if you use `safetensors` library with header inside?

**Review by alexsu52 — CHANGES_REQUESTED** [review]

File: nncf/tensor/tensor.py [hunk 10]
```diff
@@ -162,13 +164,6 @@ def __gt__(self, other: Union[Tensor, float]) -> Tensor:
     def __ge__(self, other: Union[Tensor, float]) -> Tensor:
         return Tensor(self.data >= unwrap_tensor_data(other))
 
-    # Methods to support pickling and unpickling
```
[the remark this answers is not in the document]
Comment (kshpv) [reply nncf/tensor/tensor.py:165]: It still allows using pickle with `nncf.Tensor`. Before, these methods determined what part of `Tensor` is dumped. With removing these methods pickle will dump the whole tensor structure. 

Before I introduced these methods to keep the structure of dumped tensors between versions and manually change it if needed.

For now, I think for developers is more beneficial to not fix the dumped state, so I removed it.

**Review by AlexanderDokuchaev — CHANGES_REQUESTED** [review]

File: nncf/tensor/functions/numeric.py [hunk 11]
```diff
@@ -905,3 +907,38 @@ def ceil(a: Tensor) -> Tensor:
     :return: An array of the same type as a, containing the ceiling values.
     """
     return Tensor(ceil(a.data))
+
+
+def load_file(
```
[the remark this answers is not in the document]
Comment (alexsu52) [reply nncf/tensor/functions/numeric.py:912]: https://numpy.org/doc/2.1/reference/generated/numpy.save.html - is implemented in io numpy module. This functions work with dict not with tensor. In the future the methods such as save/load should be added for tensor but this is out of scope of this PR.

File: nncf/common/tensor_statistics/aggregator.py [hunk 12]
```diff
@@ -96,16 +99,19 @@ def collect_statistics(self, model: TModel, graph: NNCFGraph) -> None:
...
         Loads statistics from a directory and populates the statistic points with the loaded data.
 
         :param dir_path: The name of the directory from which to load the statistics.
         """
-        loaded_data, metadata = statistics_serializer.load_from_dir(dir_path)
+        tensor_backend = get_tensor_backend(self.BACKEND)
+        loaded_data: Dict[str, Dict[str, Tensor]]
+        metadata: Dict[str, Any]
+        loaded_data, metadata = statistics_serializer.load_from_dir(dir_path, tensor_backend)
         statistics_validator.validate_backend(metadata, self.BACKEND)
```
[the remark this answers is not in the document]
Comment (kshpv) [reply nncf/common/tensor_statistics/aggregator.py:112]: Do we have a rule for this? I like the current version

[the remark this answers is not in the document]
Comment (kshpv) [reply nncf/common/tensor_statistics/aggregator.py:112]: Do we have a rule for this? I like the current version

File: nncf/experimental/common/tensor_statistics/collectors.py [hunk 13]
```diff
@@ -798,8 +798,8 @@ def _register_reduced_input_impl(self, x: TensorType) -> None:
         # TODO: revise this formula as possibly it is with an error; adopted from previous HAWQ implementation
         self._container = (self._container + trace) / x.size
 
-    def _aggregate_impl(self) -> List[TensorType]:
-        return [self._container * 2 / self._collected_samples]
+    def _aggregate_impl(self) -> Tensor:
```
[the remark this answers is not in the document]
Comment (kshpv) [reply nncf/experimental/common/tensor_statistics/collectors.py:801]: It was not aligned with `HessianTensorStatistic` class https://github.com/openvinotoolkit/nncf/blob/develop/nncf/experimental/common/tensor_statistics/statistics.py#L174

File: nncf/common/tensor_statistics/statistics_serializer.py [hunk 14]
```diff
@@ -8,112 +8,141 @@
...
-        except (pickle.UnpicklingError, IOError) as e:
-            raise nncf.InternalError(f"Error loading statistics from {statistics_file.name}: {e}")
-    return statistics, metadata.get("metadata", {})
+    for file_name, original_name in mapping.items():
+        statistics_file = dir_path / file_name
+        if not statistics_file.exists():
+            raise nncf.ValidationError(
+                f"No statistics file was found for {original_name}. Probably, metadata is corrupted."
+            )
+        statistics[original_name] = fns.load_file(statistics_file, backend)
```
[the remark this answers is not in the document]
Comment (kshpv) [reply nncf/common/tensor_statistics/statistics_serializer.py]: Do you mean torch tensors with device='cuda'? Yes, moreover, I added a testcase for it - https://github.com/openvinotoolkit/nncf/pull/3122/files#diff-34b3b29becfe5a7b6e31fa3c7e73e802f1ee0d4853286bb28307c2890ef5b048R1700

File: nncf/tensor/functions/numeric.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (kshpv) [reply nncf/tensor/functions/numeric.py]: added `io` module

File: nncf/experimental/common/tensor_statistics/statistics.py [hunk 15]
```diff
@@ -13,10 +13,32 @@
...
-from typing import Any, ClassVar, Dict, List, Tuple
+from typing import Any, Callable, ClassVar, Dict, List, Tuple
 
+import nncf
 from nncf.tensor import Tensor
 from nncf.tensor import functions as fns
+from nncf.tensor.definitions import TensorBackend
+
+
+def get_tensor_method(tensor_backend: TensorBackend) -> Callable:
```
[the remark this answers is not in the document]
Comment (kshpv) [reply nncf/experimental/common/tensor_statistics/statistics.py]: done

File: nncf/tensor/functions/dispatcher.py [hunk 16]
```diff
@@ -43,6 +43,26 @@ def dispatch_list(fn: "functools._SingleDispatchCallable", tensor_list: List[Ten
...
+    :param tensor_dict: Dict of Tensors.
+    :return: The result value of the function call.
+    """
+    unwrapped_dict = {}
+    tensor_backend = None
+    for key, tensor in tensor_dict.items():
+        if tensor_backend is None:
+            tensor_backend = type(tensor.data)
+        else:
+            assert tensor_backend is type(tensor.data)
```
[the remark this answers is not in the document]
Comment (kshpv) [reply nncf/tensor/functions/dispatcher.py]: done

File: nncf/experimental/common/tensor_statistics/statistics.py [hunk 17]
```diff
@@ -25,11 +47,36 @@ class TensorStatistic:
...
     def get_data(self) -> Dict[str, Any]:
+        """
+        Retrieves the data of the tensor statistics.
+
+        :return: Dictionary with keys and their associated data.
+        """
         return {key: getattr(self, key) for key in self.keys()}
 
-    def load_data(self, data: Dict[str, Any]):
+    def get_data_to_dump(self) -> Dict[str, Tensor]:
```
[the remark this answers is not in the document]
Comment (kshpv) [reply nncf/experimental/common/tensor_statistics/statistics.py]: introduced `get_data(self, is_serialized: bool = False)`. Could you let me know if it matches your request?

File: nncf/experimental/common/tensor_statistics/statistics.py [hunk 18]
```diff
@@ -25,11 +47,36 @@ class TensorStatistic:
     TENSOR_STATISTIC_OUTPUT_KEY = "tensor_statistic_output"
 
     def get_data(self) -> Dict[str, Any]:
+        """
+        Retrieves the data of the tensor statistics.
+
+        :return: Dictionary with keys and their associated data.
+        """
         return {key: getattr(self, key) for key in self.keys()}
```
[the remark this answers is not in the document]
Comment (kshpv) [reply nncf/experimental/common/tensor_statistics/statistics.py]: done

File: nncf/experimental/common/tensor_statistics/statistics.py [hunk 19]
```diff
@@ -25,11 +47,36 @@ class TensorStatistic:
...
+                serialized_data[key] = value
+            else:
+                raise nncf.InternalError(f"Unsupported type of value: {type(value)}")
+        return serialized_data
+
+    def load_data(self, loaded_data: Dict[str, Tensor]) -> None:
+        """
+        Loads the data from the serialized data.
+
+        :param: Data to load.
```
[the remark this answers is not in the document]
Comment (kshpv) [reply nncf/experimental/common/tensor_statistics/statistics.py]: done

File: nncf/common/tensor_statistics/aggregator.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (kshpv) [reply nncf/common/tensor_statistics/aggregator.py]: done

File: nncf/common/tensor_statistics/aggregator.py [hunk 20]
```diff
@@ -96,16 +99,19 @@ def collect_statistics(self, model: TModel, graph: NNCFGraph) -> None:
...
         self._load_statistics(loaded_data)
-        nncf_logger.info(f"Statistics were successfully loaded from a directory {dir_path}.")
+        nncf_logger.info(f"Statistics were successfully loaded from a directory {dir_path.absolute()}.")
```
[the remark this answers is not in the document]
Comment (kshpv) [reply nncf/common/tensor_statistics/aggregator.py]: done

File: nncf/common/tensor_statistics/statistics_serializer.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (kshpv) [reply nncf/common/tensor_statistics/statistics_serializer.py]: done

File: nncf/common/tensor_statistics/statistics_serializer.py [hunk 21]
```diff
@@ -8,112 +8,141 @@
...
         # Update the mapping
-        mapping[sanitized_name] = original_name
+        metadata["mapping"][unique_sanitized_name] = original_name
 
         try:
-            fail_if_symlink(file_path)
-            with gzip.open(file_path, "wb") as f:
-                pickle.dump(statistics_value, f)
-        except (IOError, pickle.PicklingError) as e:
+            fns.save_file(statistics_value, file_path)
```
[the remark this answers is not in the document]
Comment (kshpv) [reply nncf/common/tensor_statistics/statistics_serializer.py]: done

File: nncf/common/tensor_statistics/aggregator.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (kshpv) [reply nncf/common/tensor_statistics/aggregator.py]: done

File: nncf/common/tensor_statistics/aggregator.py [hunk 22]
```diff
@@ -142,7 +148,7 @@ def _prepare_statistics(self) -> Dict[str, Any]:
         for _, statistic_point, tensor_collector in self.statistic_points.get_tensor_collectors():
             statistics = tensor_collector.get_statistics()
             statistics_key = self._get_statistics_key(statistics, statistic_point.target_point)
-            data = statistics.get_data()
+            data: Dict[str, Tensor] = statistics.get_data_to_dump()
```
[the remark this answers is not in the document]
Comment (kshpv) [reply nncf/common/tensor_statistics/aggregator.py]: done

File: nncf/openvino/quantization/quantize_model.py [hunk 23]
```diff
@@ -403,8 +403,9 @@ def compress_weights_impl(
     statistics_points = None
     if advanced_parameters and advanced_parameters.statistics_path:
         # If there is no such directory, then caches statistics
-        if not Path(advanced_parameters.statistics_path).exists():
-            cache_weight_compression_statistics(model, graph, dataset, subset_size, advanced_parameters.statistics_path)
+        dir_path = Path(advanced_parameters.statistics_path)
```
[the remark this answers is not in the document]
Comment (kshpv) [reply nncf/openvino/quantization/quantize_model.py]: done

File: nncf/common/tensor_statistics/aggregator.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (kshpv) [reply nncf/common/tensor_statistics/aggregator.py]: done

File: tests/cross_fw/test_templates/test_statistics_serializer.py [hunk 24]
```diff
@@ -0,0 +1,116 @@
...
+
+    @abstractmethod
+    def _get_tensor_backend(self) -> TensorBackend:
+        """Returns the backend used for testing."""
+
+    @abstractmethod
+    def is_equal(self, a1: Dict[str, Tensor], a2: Dict[str, Tensor]) -> bool:
+        """Determine if two statistics are equal."""
+
+    def test_sanitize_filename(self):
```
[the remark this answers is not in the document]
Comment (kshpv) [reply tests/cross_fw/test_templates/test_statistics_serializer.py:44]: done

File: nncf/common/tensor_statistics/statistics_serializer.py (hunk 9 above)
Comment (kshpv) [reply nncf/common/tensor_statistics/statistics_serializer.py]: to store common information regarding caching: backend, subset_size. Can be extended in many ways

Comment (alexsu52) [reply nncf/common/tensor_statistics/statistics_serializer.py]: ok

**Review by alexsu52 — APPROVED** [review]
LGTM

**Review by MaximProshin — APPROVED** [review]

**Comment by MaximProshin** [comment]
sdl/job/nightly_trigger/370/

**Review by AlexanderDokuchaev — APPROVED** [review]

**Pull request merged**